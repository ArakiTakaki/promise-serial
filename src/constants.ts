export const MAX_TIMEOUT_NUMBER = 2 ** 31 - 1;
