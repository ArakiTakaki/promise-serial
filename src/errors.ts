export class CannceledError<T extends any> extends Error {
    public name: string = 'CannceledError';
    public message: string = [
        'Canceled the operation of promiseSerial.',
        'The result in the middle is stored in "error.result".',
    ].join('\n');
    public stack?: string;
    public results: T[] = [];

    constructor(results: T[]) {
        super();
        this.results = results;
    }
}
