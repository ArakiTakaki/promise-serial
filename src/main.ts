
export { CannceledError } from './errors';
export { call } from './utils';
export { 
    promiseSerial,
    PromiseSerialHandler,
    PromiseSerialOptions,
    PromiseSerialResult,
    PromiseSerialValue,
} from './promiseSerial';
export { 
    PromiseSerialMiddleware,
    PromiseSerialMiddlewareHandlers,
} from './interfaces/middleware';