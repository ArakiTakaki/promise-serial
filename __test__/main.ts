import { promiseSerial } from '../src/main';

const waitForPromise = async (wait: number, val: string): Promise<string> => {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve(val);
        }, wait)
    });
}

const main = async () => {
    {
        const results = promiseSerial([
            () => waitForPromise(100, 'done 1'),
            () => waitForPromise(100, 'done 2'),
            () => waitForPromise(100, 'done 3'),
            () => waitForPromise(100, 'done 4'),
        ])
        console.log(await results.value)
    }
    {
        try {
            const results = promiseSerial([
                () => waitForPromise(100, 'done 1'),
                () => waitForPromise(100, 'done 2'),
                () => waitForPromise(100, 'done 3'),
                () => waitForPromise(100, 'done 4'),
            ])
            console.log(await waitForPromise(110, 'cancel'));
            results.cancel();
            console.log(await results.value)
        } catch (err) {
            console.error(err);
        }
    }
}

main();
