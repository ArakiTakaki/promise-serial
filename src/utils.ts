import { MAX_TIMEOUT_NUMBER } from './constants';

export const call = <T extends (...args: any[]) => any>(cb: T, ...data: Parameters<T>) => (): ReturnType<T> => cb(...data);

export const notNull = <T>(item: T | null | undefined): item is T => {
    return item != null;
}

/**
 * イベントを呼び出し後、次のイベントまで指定した時間が経過するまではイベントを発生させない処理。
 * @param callback
 * @param wait (ms)
 */
export const debounce = <T extends (...args: any[]) => any>(func: T, wait: number = 500): {
  exec: (...args: Parameters<T>) => void,
  cancel: () => void,
} => {
  let cancelToken: NodeJS.Timeout;
  const callback = (...args: Parameters<T>) => {
    clearTimeout(cancelToken);
    cancelToken = setTimeout(() => {
      func(...args);
    }, Math.min(wait, MAX_TIMEOUT_NUMBER));
  }
  return {
    exec: callback,
    cancel: () => {
      clearTimeout(cancelToken);
    }
  };
};
