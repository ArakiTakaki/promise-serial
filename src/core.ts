import { CannceledError } from './errors';
import { PromiseSerialMiddleware } from './interfaces/middleware';

type Serializable<T extends readonly unknown[] | []> = Promise<T>;

export const serializer = async <T extends Promise<any>>(values: (() => T)[], middlewares: PromiseSerialMiddleware<T>[] = []): Serializable<T[]> => {
    const _middleware = middlewares.map(value => value());

    const main = async () => {
        const results: T[] = [];

        const cancelProcess = (err: Error) => {
            _middleware.forEach((value) => value.error({
                values,
                results,
                errorObject: err,
            }));
            _middleware.forEach((value) => value.finished({
                results,
            }));
            
            if (err != null) throw new CannceledError<T>(results);
        }

        for (let i = 0; i < values.length; i ++ ) {
            // before updated
            _middleware.forEach(value => value.beforeUpdate({
                values: values,
                results: results,
                index: i,
                throws: cancelProcess,
            }));

            const result = await values[i]();
            // update
            _middleware.forEach(value => value.update({
                values: values,
                results: results,
                index: i,
                throws: cancelProcess,
            }));
            results.push(result);
            _middleware.forEach(value => value.updated({
                values: values,
                results: results,
                index: i,
                throws: cancelProcess,
            }));
        }
        _middleware.forEach(value => value.finished({
            results: results,
        }));
        return results;
    };
    const process = main()
    return await _middleware.reduce((process, { editResult }) => {
        if (editResult == null) return process;
        return editResult({
            process: process,
        });
    }, process);
};