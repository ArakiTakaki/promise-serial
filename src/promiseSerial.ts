import { serializer } from './core';
import { PromiseSerialMiddleware } from './main';
import { cancelMiddleware } from './middlewares/cancel';
import { ProgressHandler, progressMiddleware } from './middlewares/progress';
import { notNull } from './utils';

export type PromiseSerialValue<T> = (() => T)[];
export interface PromiseSerialResult<T extends readonly unknown[] | []> {
    value: Promise<T>;
    cancel: () => Promise<T>;
}

export interface PromiseSerialOptions<T> {
    onProgress?: ProgressHandler<T>,
    timeout?: number;
    isNotCancelledThrow?: boolean;
    middlewares?: PromiseSerialMiddleware<T>[];
}

export type PromiseSerialHandler<T extends Promise<any>> = (values: PromiseSerialValue<T>, options: PromiseSerialOptions<T>) => PromiseSerialResult<T>;

/**
 * Promiseを直列実行させるための関数
 * @param values 同期処理をするPromise群
 * @param options.onProgress 進行状況を返却するコールバックの定義
 * @param options.timeout 各イベントのタイムアウト期間 (初期値：無限)
 * @param options.isNotCancelledThrow canncel時にthrowするかどうか
 * @returns result.value Promiseを返却
 * @returns result.cancel() 実行時にキャンセル
 * @returns result.progress() 現在の進捗 0-1
 */
export const promiseSerial = <T extends Promise<any>>(values: PromiseSerialValue<T>, options: PromiseSerialOptions<T> = {}): PromiseSerialResult<T[]> => {
    const cancellable = cancelMiddleware<T>(options.timeout, options.isNotCancelledThrow);
    const middlewares = [
        cancellable.middleware,
        options.onProgress != null ? progressMiddleware(options.onProgress) : undefined,
        ...(options.middlewares || []),
    ].filter(notNull)

    const process = serializer<T>(values, middlewares);

    return {
        value: process,
        cancel: () => {
            cancellable.cancel();
            return process;
        }
    }
};
