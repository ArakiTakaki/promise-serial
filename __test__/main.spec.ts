import { promiseSerial } from '../src/main';
import { call } from '../src/utils';

const waitForTest = (returnValue: string) => async (waitTime: number): Promise<string> => {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve(returnValue);
        }, waitTime);
    });
};

describe('promise serial', () => {
    it('basic', async () => {
        const values = ['a', 'b', 'c', 'd', 'e', 'f'];
        const result = await promiseSerial(values.map(waitForTest).map((cb, index) => call(cb, 100 * index)), {
            onProgress: (progress, index) => {
                const progressValue = 1 / values.length;
                const containResult = new Array(values.length).fill(0).map((_, index) => {
                    return (index + 1) * progressValue;
                });
                expect(progress).toBeCloseTo(containResult[index]);
            },
        }).value;
        expect(result).toEqual(values);
    });

    it('timeout', (done) => {
        const main = async () => {
            const values = ['a', 'b', 'c', 'd', 'e', 'f'];
            const result = promiseSerial(values.map(waitForTest).map((cb, index) => call(cb, 100 * index)), {
                timeout: 200,
            });
            await expect(result.value).rejects.toThrowErrorMatchingSnapshot();
            done();
        }
        main();
    });

    it('cancel', (done) => {
        const exec = async () => {
            const values = ['a', 'b', 'c', 'd', 'e', 'f'];
            const result = promiseSerial(values.map(waitForTest).map((cb, index) => call(cb, 100 * index)), {
                timeout: Infinity,
            });
            setTimeout(() => {
                result.cancel();
            }, 200)
            await expect(result.value).rejects.toThrowErrorMatchingSnapshot();
            done();
        }
        exec();
    });

    it('Cancel not throw', async () => {
        const values = ['a', 'b', 'c', 'd', 'e', 'f'];
        const answer = ['a'];
        const result = promiseSerial(values.map(waitForTest).map((cb) => call(cb, 100)), {
            timeout: Infinity,
            isNotCanceledThrow: true,
        });

        setTimeout(() => {
            expect(result.cancel()).resolves.toEqual(answer);
        }, 120);

        await expect(result.value).resolves.toEqual(answer);
    });
});

describe('issue', () => {
    it('#17', async () => {
        const values = ['a', 'b', 'c', 'd', 'e', 'f'];
        const result = await promiseSerial(values.map(waitForTest).map((cb, index) => call(cb, 100 * index))).value;
        expect(typeof result[0]).toEqual('string');
    });
});