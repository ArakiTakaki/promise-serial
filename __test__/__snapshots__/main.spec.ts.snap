// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`promise serial cancel 1`] = `
"Canceled the operation of promiseSerial.
The result in the middle is stored in \\"error.result\\"."
`;

exports[`promise serial timeout 1`] = `
"Canceled the operation of promiseSerial.
The result in the middle is stored in \\"error.result\\"."
`;
